### PR TITLE
xd: init at 0.4.0

### LINKED ISSUE #93055: Package request: XD (i2p)
**Project description**
> I2P BitTorrent Client written in GO

* homepage URL: https://xd-torrent.github.io/
* source URL: https://github.com/majestrate/XD
* license: mit
* platforms: linux

### PR DESCRIPTION
**Pull request by nixbitcoin**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Includes script for fetching GPG verified sha256.

Module may follow and is already working @ https://github.com/nixbitcoin/nix-darknet/blob/master/modules/XD.nix

Closes https://github.com/NixOS/nixpkgs/issues/93055

### CODE REVIEW

File: pkgs/applications/networking/p2p/xd/default.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  postInstall = ''
+    ln -s $out/bin/XD $out/bin/XD-CLI
+  '';
+
+  meta = with lib; {
+    description = ''
+      i2p bittorrent client
+    '';
+    homepage = "https://xd-torrent.github.io";
+    maintainers = with maintainers; [ nixbitcoin ];
```
Comment (NULLx76) [inline pkgs/applications/networking/p2p/xd/default.nix:26]: It seems like your forgot to add yourself to `nixpkgs/maintainers/maintainer-list.nix` so `nixbitcoin` is undefined here. See the [manual](https://nixos.org/manual/nixpkgs/unstable/#var-meta-maintainers) for the specifics.

Here is the full error for completeness' sake.
image

File: pkgs/applications/networking/p2p/xd/get-sha256.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+export GNUPGHOME=$TMPDIR
+echo "Fetching Jeff Becker's key"
+gpg --keyserver hkps://keyserver.ubuntu.com --recv-keys 67EF6BA68E7B0B0D6EB4F7D4F357B3B42F6F9B05 2> /dev/null
+
+echo "Verifying latest release"
+git verify-tag ${latest}
+
+echo "tag: ${latest}"
+# The prefix option is necessary because GitHub prefixes the archive contents in this format
+echo "sha256: $(git archive --format tar.gz --prefix=XD-${latest//v}/ ${latest} | sha256sum | cut -d\  -f1)"
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/xd/get-sha256.sh:25]: Please remove this and just use nix-update.

File: pkgs/applications/networking/p2p/xd/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+buildGoModule rec {
+  pname = "XD";
+  version = "0.4.0";
+
+  src = fetchurl {
+    url = "https://github.com/majestrate/XD/archive/v${version}.tar.gz";
+    # Use ./get-sha256.sh to fetch latest (verified) sha256
+    sha256 = "18faab1bae26721feb60487c190dda07c7297a9d35d69531074fbd683a3bbc02";
+  };
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/xd/default.nix:11]: Please use fetchFromGitHub.

File: pkgs/applications/networking/p2p/xd/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  '';
+
+  meta = with lib; {
+    description = ''
+      i2p bittorrent client
+    '';
+    homepage = "https://xd-torrent.github.io";
+    maintainers = with maintainers; [ nixbitcoin ];
+    license = lib.licenses.mit;
+    platforms = lib.platforms.linux;
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/xd/default.nix:28]:
```suggestion
    license = licenses.mit;
    platforms = platforms.linux;
```

File: pkgs/applications/networking/p2p/xd/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  nativeBuildInputs = [ perl ];
+
+  postInstall = ''
+    ln -s $out/bin/XD $out/bin/XD-CLI
+  '';
+
+  meta = with lib; {
+    description = ''
+      i2p bittorrent client
+    '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/p2p/xd/default.nix:24]:
```suggestion
    description = "i2p bittorrent client";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/p2p/xd/default.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  src = fetchurl {
+    url = "https://github.com/majestrate/XD/archive/v${version}.tar.gz";
+    # Use ./get-sha256.sh to fetch latest (verified) sha256
+    sha256 = "18faab1bae26721feb60487c190dda07c7297a9d35d69531074fbd683a3bbc02";
+  };
+
+  vendorSha256 = "1wg3cym2rwrhjsqlgd38l8mdq5alccz808465117n3vyga9m35lq";
+
+  nativeBuildInputs = [ perl ];
```
Comment (ThibautMarty) [inline pkgs/applications/networking/p2p/xd/default.nix:15]:
```suggestion
  checkInputs = [ perl ];
```

Perl is used in tests to provide the `shasum` executable.

File: pkgs/applications/networking/p2p/xd/default.nix (hunk 5 above)
Comment (nixbitcoin) [reply pkgs/applications/networking/p2p/xd/default.nix]: Fixed in 7a64336

File: pkgs/applications/networking/p2p/xd/default.nix (hunk 4 above)
Comment (nixbitcoin) [reply pkgs/applications/networking/p2p/xd/default.nix]: Fixed in 7a64336

File: pkgs/applications/networking/p2p/xd/get-sha256.sh (hunk 2 above)
Comment (nixbitcoin) [reply pkgs/applications/networking/p2p/xd/get-sha256.sh]: Fixed in 7a64336

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix [hunk 7]
```diff
@@ -7201,6 +7201,16 @@
     githubId = 16385648;
     name = "Niko Pavlinek";
   };
+  nixbitcoin = {
+    email = "nixbitcoin@i2pmail.org";
+    github = "nixbitcoin";
+    githubID = 45737139;
```
[the remark this answers is not in the document]
Comment (nixbitcoin) [reply maintainers/maintainer-list.nix]: Fixed in 542e957965dd12bdf54ff9417a78b6dbfea616e7

**Pull request merged**